### PR TITLE
Implementation Optimal strategy for a game (set-1) by dynamic program…(#1622)

### PR DESCRIPTION
**Pull request by biekymahato**
The issue number is #1622 for these PRs.

#### Checklist:
- [x] 4 space indentation.
- [x] Coding conventions are followed.
- [x] Input is taken dynamically.
- [x] Sample Input / Output is added at the end of file.
- [x] Logic Documentation (Comments).
- [x] File names are correct.

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

#### Languages Used:
- Cpp

#### Files Added:
- Optimal Strategy for a Game (set 1) By dynamic Programming.
- The Given solution can be optimized by using less number of comparisons for every choice.
- Added three different test cases for better understanding.

> We're happy to help you get this ready -- don't be afraid to ask for help.

Thanks!

### CODE REVIEW

**Comment by biekymahato** [comment]
@jainaman224 @asha15  @SakshiShreya @somya-kapoor 
please can you review it?

File: Optimal Strategy for a game (Set -1)/optimal_strategy_for_a_game_set_1.cpp [hunk 1]
```diff
@@ -0,0 +1,113 @@
+/*
+OPTIMAL STRATEGY FOR A GAME SET - 1
+Given Problem Statement :
+             Problem statement: Consider a row of n coins of values v1 . . . vn, where n is even.
+             We play a game against an opponent by alternating turns. In each turn,
+             a player selects either the first or last coin from the row,
+             removes it from the row permanently, and receives the value of the coin.
+             Determine the maximum possible amount of money we can definitely win if we move first.
```
Comment (asha15) [inline Optimal Strategy for a game (Set -1)/optimal_strategy_for_a_game_set_1.cpp]: keep only 4 space before the description

File: Optimal Strategy for a game (Set -1)/optimal_strategy_for_a_game_set_1.cpp [hunk 2]
```diff
@@ -0,0 +1,113 @@
+/*
+OPTIMAL STRATEGY FOR A GAME SET - 1
+Given Problem Statement :
+             Problem statement: Consider a row of n coins of values v1 . . . vn, where n is even.
+             We play a game against an opponent by alternating turns. In each turn,
+             a player selects either the first or last coin from the row,
+             removes it from the row permanently, and receives the value of the coin.
```
Comment (asha15) [inline Optimal Strategy for a game (Set -1)/optimal_strategy_for_a_game_set_1.cpp]: remove the "," before the word "and"

File: Optimal Strategy for a game (Set -1)/optimal_strategy_for_a_game_set_1.cpp [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+             Determine the maximum possible amount of money we can definitely win if we move first.
+       Note - Note: The opponent is as clever as the user.
+*/
+
+
+
+
```
Comment (asha15) [inline Optimal Strategy for a game (Set -1)/optimal_strategy_for_a_game_set_1.cpp]: remove these lines

File: Optimal Strategy for a game (Set -1)/optimal_strategy_for_a_game_set_1.cpp [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+// C++ program to find out maximum value from a given sequence of coins
+  #include <bits/stdc++.h>
+  using namespace std;
```
Comment (asha15) [inline Optimal Strategy for a game (Set -1)/optimal_strategy_for_a_game_set_1.cpp]: deleted the space before #include and using

File: Optimal Strategy for a game (Set -1)/optimal_strategy_for_a_game_set_1.cpp [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+// Returns optimal value possible that a player can
+// collect from an array of coins of size n. Note
+// than n must be even
```
Comment (asha15) [inline Optimal Strategy for a game (Set -1)/optimal_strategy_for_a_game_set_1.cpp]: make this as a multi line comment

File: Optimal Strategy for a game (Set -1)/optimal_strategy_for_a_game_set_1.cpp [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+int optimalStrategyOfGame(int* arr, int n)
+{
+    // Create a table to store solutions of subproblems
+    int table[n][n];
+
+    // Fill table using above recursive formula. Note
+    // that the table is filled in diagonal fashion from diagonal elements to table[0][n-1] which is the result.
```
Comment (asha15) [inline Optimal Strategy for a game (Set -1)/optimal_strategy_for_a_game_set_1.cpp]: change this also as a multi line commend

File: Optimal Strategy for a game (Set -1)/optimal_strategy_for_a_game_set_1.cpp [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+    for (int gap = 0; gap < n; ++gap) {
+        for (int i = 0, j = gap; j < n; ++i, ++j) {
+
```
Comment (asha15) [inline Optimal Strategy for a game (Set -1)/optimal_strategy_for_a_game_set_1.cpp]: Remove this line

File: Optimal Strategy for a game (Set -1)/optimal_strategy_for_a_game_set_1.cpp [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+            // Here x is value of F(i+2, j), y is F(i+1, j-1) and
```
Comment (asha15) [inline Optimal Strategy for a game (Set -1)/optimal_strategy_for_a_game_set_1.cpp]: keep space before and after arithmetic operators, also make this as a multi line commend

File: Optimal Strategy for a game (Set -1)/optimal_strategy_for_a_game_set_1.cpp [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+            int y = ((i + 1) <= (j - 1)) ? table[i + 1][j - 1] : 0;
+            int z = (i <= (j - 2)) ? table[i][j - 2] : 0;
+
+            table[i][j] = max(arr[i] + min(x, y), arr[j] + min(y, z));
+        }
+    }
+
+    return table[0][n - 1];
+}
+
```
Comment (asha15) [inline Optimal Strategy for a game (Set -1)/optimal_strategy_for_a_game_set_1.cpp]: remove this line

File: Optimal Strategy for a game (Set -1)/optimal_strategy_for_a_game_set_1.cpp [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+
+// Driver Code to test above function
+
```
Comment (asha15) [inline Optimal Strategy for a game (Set -1)/optimal_strategy_for_a_game_set_1.cpp]: remove this line too

File: Optimal Strategy for a game (Set -1)/optimal_strategy_for_a_game_set_1.cpp [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+int main()
+{
+
```
Comment (asha15) [inline Optimal Strategy for a game (Set -1)/optimal_strategy_for_a_game_set_1.cpp]: remove this line

File: Optimal Strategy for a game (Set -1)/optimal_strategy_for_a_game_set_1.cpp [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+    int arr1[1005],n;
```
Comment (asha15) [inline Optimal Strategy for a game (Set -1)/optimal_strategy_for_a_game_set_1.cpp]: keep a space after comma

File: Optimal Strategy for a game (Set -1)/optimal_strategy_for_a_game_set_1.cpp [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+
```
Comment (asha15) [inline Optimal Strategy for a game (Set -1)/optimal_strategy_for_a_game_set_1.cpp]: remove this line

File: Optimal Strategy for a game (Set -1)/optimal_strategy_for_a_game_set_1.cpp [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+    cin>>n;                                //Number of values in array you want to store
```
Comment (asha15) [inline Optimal Strategy for a game (Set -1)/optimal_strategy_for_a_game_set_1.cpp]: - move the commend into previous line 
- keep space before and after >>

File: Optimal Strategy for a game (Set -1)/optimal_strategy_for_a_game_set_1.cpp [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+
+    for(int i =0; i < n; i++)              //Input taken in array
```
Comment (asha15) [inline Optimal Strategy for a game (Set -1)/optimal_strategy_for_a_game_set_1.cpp]: move the comment into previous line

File: Optimal Strategy for a game (Set -1)/optimal_strategy_for_a_game_set_1.cpp [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+         cin>>arr1[i];
```
Comment (asha15) [inline Optimal Strategy for a game (Set -1)/optimal_strategy_for_a_game_set_1.cpp]: keep space before and after >>

File: Optimal Strategy for a game (Set -1)/optimal_strategy_for_a_game_set_1.cpp [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+
+    printf("%d\n", optimalStrategyOfGame(arr1, n));          //function call for calculating it's optimal strategy
+
```
Comment (asha15) [inline Optimal Strategy for a game (Set -1)/optimal_strategy_for_a_game_set_1.cpp]: remove this line

File: Optimal Strategy for a game (Set -1)/optimal_strategy_for_a_game_set_1.cpp [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+{
+
+    int arr1[1005],n;
+
+    cin>>n;                                //Number of values in array you want to store
+
+    for(int i =0; i < n; i++)              //Input taken in array
+         cin>>arr1[i];
+
+    printf("%d\n", optimalStrategyOfGame(arr1, n));          //function call for calculating it's optimal strategy
```
Comment (asha15) [inline Optimal Strategy for a game (Set -1)/optimal_strategy_for_a_game_set_1.cpp]: move the commend to the previous line

File: Optimal Strategy for a game (Set -1)/optimal_strategy_for_a_game_set_1.cpp [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+}
+
+
+
+
+
+
+
+
+
```
Comment (asha15) [inline Optimal Strategy for a game (Set -1)/optimal_strategy_for_a_game_set_1.cpp]: Remove this unnecessary lines

**Comment by biekymahato** [comment]
@asha15  can you have a look? I change the request.

File: Optimal_Strategy_For_A_Game (Set-1)/optimal_strategy_for_a_game_set_1.cpp [hunk 20]
```diff
@@ -0,0 +1,86 @@
...
+*/
+
+// C++ program to find out maximum value from a given sequence of coins
+#include <bits/stdc++.h>
+using namespace std;
+
+/*Returns optimal value possible that a player can
+collect from an array of coins of size n. Note
+than n must be even*/
+
```
Comment (asha15) [inline Optimal_Strategy_For_A_Game (Set-1)/optimal_strategy_for_a_game_set_1.cpp]: remove this line

File: Optimal_Strategy_For_A_Game (Set-1)/optimal_strategy_for_a_game_set_1.cpp [hunk 21]
```diff
@@ -0,0 +1,86 @@
...
+int optimalStrategyOfGame(int* arr, int n)
+{
+// Create a table to store solutions of subproblems
+    int table[n][n];
+/*Fill table using above recursive formula. Note
```
Comment (asha15) [inline Optimal_Strategy_For_A_Game (Set-1)/optimal_strategy_for_a_game_set_1.cpp]: keep a line space before this commend

File: Optimal_Strategy_For_A_Game (Set-1)/optimal_strategy_for_a_game_set_1.cpp [hunk 22]
```diff
@@ -0,0 +1,86 @@
...
+that the table is filled in diagonal fashion from diagonal elements to table[0][n-1] which is the result.*/
```
Comment (asha15) [inline Optimal_Strategy_For_A_Game (Set-1)/optimal_strategy_for_a_game_set_1.cpp]: break this into two line

File: Optimal_Strategy_For_A_Game (Set-1)/optimal_strategy_for_a_game_set_1.cpp [hunk 23]
```diff
@@ -0,0 +1,86 @@
...
+    for (int gap = 0; gap < n; ++gap)
+       {
```
Comment (asha15) [inline Optimal_Strategy_For_A_Game (Set-1)/optimal_strategy_for_a_game_set_1.cpp]: why did you keep space before '{'? remove it . make the correct indentation

File: Optimal_Strategy_For_A_Game (Set-1)/optimal_strategy_for_a_game_set_1.cpp [hunk 24]
```diff
@@ -0,0 +1,86 @@
...
+         for (int i = 0, j = gap; j < n; ++i, ++j)
+           {
```
Comment (asha15) [inline Optimal_Strategy_For_A_Game (Set-1)/optimal_strategy_for_a_game_set_1.cpp]: remove the extra space before open brace

File: Optimal_Strategy_For_A_Game (Set-1)/optimal_strategy_for_a_game_set_1.cpp [hunk 25]
```diff
@@ -0,0 +1,86 @@
...
+/*Here x is value of F(i+2, j), y is F(i+1, j-1) and
+z is F(i, j-2) in above recursive formula*/
```
Comment (asha15) [inline Optimal_Strategy_For_A_Game (Set-1)/optimal_strategy_for_a_game_set_1.cpp]: keep 4 space before the commend

File: Optimal_Strategy_For_A_Game (Set-1)/optimal_strategy_for_a_game_set_1.cpp [hunk 26]
```diff
@@ -0,0 +1,86 @@
...
+
+int optimalStrategyOfGame(int* arr, int n)
+{
+// Create a table to store solutions of subproblems
+    int table[n][n];
+/*Fill table using above recursive formula. Note
+that the table is filled in diagonal fashion from diagonal elements to table[0][n-1] which is the result.*/
+    for (int gap = 0; gap < n; ++gap)
+       {
+         for (int i = 0, j = gap; j < n; ++i, ++j)
```
Comment (asha15) [inline Optimal_Strategy_For_A_Game (Set-1)/optimal_strategy_for_a_game_set_1.cpp]: remove the space before parenthesis

File: Optimal_Strategy_For_A_Game (Set-1)/optimal_strategy_for_a_game_set_1.cpp [hunk 27]
```diff
@@ -0,0 +1,86 @@
...
+           {
+/*Here x is value of F(i+2, j), y is F(i+1, j-1) and
+z is F(i, j-2) in above recursive formula*/
+               int x = ((i + 2) <= j) ? table[i + 2][j] : 0;
+               int y = ((i + 1) <= (j - 1)) ? table[i + 1][j - 1] : 0;
+               int z = (i <= (j - 2)) ? table[i][j - 2] : 0;
+               table[i][j] = max(arr[i] + min(x, y), arr[j] + min(y, z));
+            }
+         }
+return table[0][n - 1];
```
Comment (asha15) [inline Optimal_Strategy_For_A_Game (Set-1)/optimal_strategy_for_a_game_set_1.cpp]: correct the indentation

File: Optimal_Strategy_For_A_Game (Set-1)/optimal_strategy_for_a_game_set_1.cpp [hunk 28]
```diff
@@ -0,0 +1,86 @@
...
+     //Number of values in array you want to store
+     cin >> n;
+     //Input taken in array
+     for(int i = 0 ; i < n ; i ++ )
+             cin >> arr1[i];
+
+    //function call for calculating it's optimal strategy
+    printf("%d\n", optimalStrategyOfGame(arr1, n));
+
+return 0;
```
Comment (asha15) [inline Optimal_Strategy_For_A_Game (Set-1)/optimal_strategy_for_a_game_set_1.cpp]: correct the indentation (keep four spaces )

File: Optimal_Strategy_For_A_Game (Set-1)/optimal_strategy_for_a_game_set_1.cpp [hunk 29]
```diff
@@ -0,0 +1,86 @@
...
+}
```
Comment (asha15) [inline Optimal_Strategy_For_A_Game (Set-1)/optimal_strategy_for_a_game_set_1.cpp]: keep a line space after this

**Comment by biekymahato** [comment]
@asha15  done request-2 changes have a look.

File: Optimal_Strategy_For_A_Game (Set-1)/optimal_strategy_for_a_game_set_1.cpp [hunk 30]
```diff
@@ -0,0 +1,90 @@
...
+      for (int gap = 0; gap < n; ++gap){
+         for (int i = 0, j = gap; j < n; ++i, ++j){
+
+    /*Here x is value of F(i+2, j), y is F(i+1, j-1) and
+    z is F(i, j-2) in above recursive formula*/
+             int x = ((i + 2) <= j) ? table[i + 2][j] : 0;
+             int y = ((i + 1) <= (j - 1)) ? table[i + 1][j - 1] : 0;
+             int z = (i <= (j - 2)) ? table[i][j - 2] : 0;
+             table[i][j] = max(arr[i] + min(x, y), arr[j] + min(y, z));
+         }
```
Comment (asha15) [inline Optimal_Strategy_For_A_Game (Set-1)/optimal_strategy_for_a_game_set_1.cpp]: please correct the indentation @biekymahato , please keep spaces

**Pull request merged**